tools/checklocks/analysis.go

```diff
@@ -497,35 +497,21 @@ func (pc *passContext) checkInstruction(inst ssa.Instruction, ls *lockState) (*s
 	return nil, nil
 }
 
-// checkBasicBlock traverses the control flow graph starting at a set of given
-// block and checks each instruction for allowed operations.
-func (pc *passContext) checkBasicBlock(fn *ssa.Function, block *ssa.BasicBlock, lff *lockFunctionFacts, parent *lockState, seen map[*ssa.BasicBlock]*lockState) *lockState {
-	if oldLS, ok := seen[block]; ok && oldLS.isCompatible(parent) {
-		return nil
-	}
-
-	// If the lock state is not compatible, then we need to do the
-	// recursive analysis to ensure that it is still sane. For example, the
-	// following is guaranteed to generate incompatible locking states:
-	//
-	//	if foo {
-	//		mu.Lock()
-	//	}
-	//	other stuff ...
-	//	if foo {
-	//		mu.Unlock()
-	//	}
-
+// checkBasicBlock checks each instruction of basic block for allowed operations.
+//
+// ls is the post lock state after the final instruction in the block is checked.
+//
+// returns is true if the basic block has a return statement as the last instruction.
+func (pc *passContext) checkBasicBlock(fn *ssa.Function, block *ssa.BasicBlock, lff *lockFunctionFacts, entry *lockState) (post *lockState, returns bool) {
```

Comment:
I try to avoid named return values unless it improves readability, e.g. I would reference it in a comment. (Or you need to bind it in a lambda to handle a recover() in a defer.)

(Often naming return values is a net readability loss IMO. More code. The return statement has a different arity. More variables to track.)